### PR TITLE
Implement API methods to create and remove data sets

### PR DESCRIPTION
**Pull request by jjmerchante**
This Pull request includes the following changes:
- The implementation didn't allow to have Projects without Ecosystems but it models it was allowed. Remove in this pull requests the `null=True` attribute in Ecosystem and fix some tests that weren't using a default Ecosystem.
- New API method to add datasets as it was defined in https://github.com/chaoss/grimoirelab-bestiary/issues/39
- New API method to remove datasets by id
- GraphQL query to get dataset objects. It includes filters and pagination methods.
- GraphQL mutation to delete a dataset object by ID.

Every change has its related tests.

### CODE REVIEW

File: tests/test_api.py [hunk 1]
```diff
@@ -968,6 +1246,138 @@ def test_operations(self):
...
+            api.delete_dataset(self.ctx, dataset_id=None)
+
+        # Check if there are no transactions created when there is an error
+        transactions = Transaction.objects.filter(created_at__gt=trx_date)
+        self.assertEqual(len(transactions), 0)
+
+    def test_remove_datasource(self):
+        """Check if removing all data sets related with a data source deletes the data source"""
+        api.delete_dataset(self.ctx, dataset_id=self.dataset_1.id)
+        api.delete_dataset(self.ctx, dataset_id=self.dataset_2.id)
```
Comment (sduenas) [inline tests/test_api.py]: Leave an empty line between the code and the docstring.

File: tests/test_api.py [hunk 2]
```diff
@@ -745,6 +755,274 @@ def test_operations(self):
...
+                            'issues',
+                            None)
+
+        # Check if there are no transactions created when there is an error
+        transactions = Transaction.objects.all()
+        self.assertEqual(len(transactions), 0)
+
+    def test_unique_datasource_created(self):
+        """Check if only a data source is created when many data sets are added"""
+        dataset_1 = api.add_dataset(self.ctx,
```
Comment (sduenas) [inline tests/test_api.py]: Leave an empty line between the code and the docstring.

File: bestiary/core/schema.py [hunk 3]
```diff
@@ -233,6 +256,16 @@ class ProjectFilterType(graphene.InputObjectType):
...
 
 
+class DatasetFilterType(graphene.InputObjectType):
+    """Fields which can be used as a filter for DatasetPaginatedType objects"""
+
+    id = graphene.ID(required=False)
+    project_id = graphene.ID(required=False)
+    datasource_name = graphene.String(required=False)
+    uri = graphene.String(required=False)
+    category = graphene.String(required=False)
```
Comment (sduenas) [inline bestiary/core/schema.py]: Can you add some documentation about these fields? You can check in sortinghat how to do it: https://github.com/chaoss/grimoirelab-sortinghat/blob/muggle/sortinghat/core/schema.py#L275

**Review by sduenas — CHANGES_REQUESTED** [review]
In overall it looks good to me. There are some comments to be addressed, though.

**Comment by jjmerchante** [comment]
I have a question about the JSONField. I have included some tests that now fail, but I don't understand why.

As far as I understand there is a function to serialize the JSONFields and it is registered with `@convert_django_field.register(JSONField)`. It works for `Operation.args` but it doesn't work for `Dataset.filters` by default. I don't know what I'm missing.

**Comment by sduenas** [comment]
> I have a question about the JSONField. I have included some tests that now fail, but I don't understand why.
> 
> As far as I understand there is a function to serialize the JSONFields and it is registered with `@convert_django_field.register(JSONField)`. It works for `Operation.args` but it doesn't work for `Dataset.filters` by default. I don't know what I'm missing.

What's the error you are getting?

**Comment by jjmerchante** [comment]
Debugging, the error is returned here:
```
executed = client.execute(BT_DATASETS_QUERY,
                          context_value=self.context_value)
datasets = executed['data']['datasets']['entities']
```
```
{'errors': [{'message': 'the JSON object must be str, bytes or bytearray, not dict'}], 'data': {'datasets': {'entities': [None]}}}
```
and debugging a bit more it raises here: https://github.com/chaoss/grimoirelab-bestiary/blob/unicorn/bestiary/core/schema.py#L99

The value is a dictionary and it expects an string. I see that in Operations the JSONField is stored as a string instead of a dictionary: https://github.com/chaoss/grimoirelab-bestiary/blob/3e26e65583ab65ccddf795d52b290922bae5985a/bestiary/core/log.py#L139, so I guess I should do the same for the Dataset (store filters as string in the JSONField)

**Review by sduenas — APPROVED** [review]
LGTM

**Review by sduenas — CHANGES_REQUESTED** [review]
The code looks ok but I forgot something. Can you add a changelog entry to keep track of the changes. See this file for more info: https://github.com/chaoss/grimoirelab-sortinghat/blob/muggle/CONTRIBUTING.md . You have also some examples in: https://github.com/chaoss/grimoirelab-bestiary/tree/unicorn/releases/unreleased

**Review by sduenas — APPROVED** [review]
LGTM

**Pull request merged**